File: LeadToOffer/19_PrintMatrixClockWise.cpp
// Little bit hard. Consider all the conditions.
class Solution {
public:
    void printMatrixInCircle(int start, int rows, int cols, vector<int> &res, vector<vector<int> > matrix) {
        int end_col = cols - 1 - start;
        int end_row = rows - 1 - start;

        for (int i=start; i<=end_col; i++) {
            res.push_back(matrix[start][i]);
        }

        for (int i=start+1; i<=end_row; i++) {
            res.push_back(matrix[i][end_col]);
        }

        if (end_row>start){
            for (int i=end_col-1; i>=start; i--) {
                res.push_back(matrix[end_row][i]);
            }
        }

        if (end_col>start) {
            for (int i= end_row-1; i>=start+1; i--) {
                res.push_back(matrix[i][start]);
            }
        }

        return;
    }

    vector<int> printMatrix(vector<vector<int> > matrix) {
        vector<int> res;
        if (matrix.size()==0 || matrix[0].size()==0)
            return res;

        int rows = matrix.size();
        int cols = matrix[0].size();
        int start = 0;

        while (start*2 < rows && start*2 < cols) {
            printMatrixInCircle(start, rows, cols, res, matrix);
            start++;
        }
        return res;
    }
};

File: LeadToOffer/19_PrintMatrixClockWise.py
## Little bit hard. Consider all the conditions.
class Solution:
    def printMatrix(self, matrix):
        res = []
        if len(matrix)==0 or len(matrix[0])==0:
            return res
        start = 0
        row = len(matrix)
        col = len(matrix[0])
        while (2*start < row and 2*start <col):
            res.extend(self.printMatrixCycle(start, row, col, matrix))
            start += 1
        return res

    def printMatrixCycle(self, start, row, col, matrix):
        res = []
        end_col = col-start
        end_row = row-start

        for i in range(start, end_col):
            res.append(matrix[start][i])
        for i in range(start+1, end_row):
            res.append(matrix[i][end_col-1])
        if end_row-1 > start:
            for i in range(end_col-2, start-1, -1):
                res.append(matrix[end_row-1][i])
        if end_col-1 > start:
            for i in range(end_row-2, start, -1):  ## for range with neg stride, need to specify it.
                res.append(matrix[i][start])
        return res
